### PR TITLE
Docker!

### PR DESCRIPTION
**Pull request by bfirsh**
This was prompted by a conversation with @jaylett the other day. Spacelog would survive much longer if we could put the whole thing in a Docker container and put it on some hosted service that runs Docker containers (Heroku?).

It's got to the point where it's sort of working. Here's how to run it:

    docker-compose run website make reindex
    docker-compose up

Then, it's running at http://dev.spacelog.org:8001/

Some fiddling was required to make this work, which I've documented in the commits.

## Design

The Redis and Xapian databases are relatively small and are generated from data in Git, so perhaps we can just bake them directly into the Docker image, instead of managing a live Redis server and migrating it. Sort of [flat data](https://next.github.com/projects/flat-data/) style where a change to the database == a deploy.

We could use [redislite](https://redislite.readthedocs.io/en/latest/) to embed redis directly into Python, but this might use up too much memory (DB is ~50MB IIRC, multiplied by number of gunicorn processes?). A simpler solution might be to just use [some init system](https://ahmet.im/blog/minimal-init-process-for-containers/) to run gunicorn and redis alongside each other.

## Todo

- [x] Static assets (both development and production aren't working yet, I think)
- [x] Index at build time then bake the redis database and server into the Docker image. That way we don't have to worry about running a separate redis server, a deploy of new data is just deploying a new Docker image, etcetc.
- [x] Make a production image.
- [x] Deploy it somewhere.

### CODE REVIEW

File: backend/indexer.py [hunk 1]
```diff
@@ -629,8 +629,8 @@ def index_meta(self):
         new_db = current_db
         print "Reindexing into database %s" % new_db
         print "Note that this is not perfect! Do not use in production."
-        redis_conn.select(new_db)
-        redis_conn.set("hold", "1")
+        # redis_conn.select(new_db)
+        # redis_conn.set("hold", "1")
```
Comment (jaylett) [inline backend/indexer.py:633]: I'd argue that with dockerisation there is zero reason to have the db flipping mechanism and we should just drop db selection, the magic key that tells us what the live database is, and so remove `flip_db` entirely.

Comment (bfirsh) [reply backend/indexer.py:633]: Makes sense. I suppose we can always revert it if we want to bring back the old deployment system.

I wonder what else we can delete? 😈

**Comment by SteveMarshall** [comment]
I’m gonna try to have a look at this over the next couple of days, and will see if I can work out the asset stuff 🤔

**Comment by SteveMarshall** [comment]
@bfirsh I've got it all working, and will have a look over the next couple of days. A couple of notes so far, though…

- The first time I tried to run `docker-compose run website make reindex`, it failed. The error is fairly long, so I've included it below. I can't reproduce it now, though, despite removing the containers, images, and volumes that were generated.
- It also failed a second time (which I can reproduce consistently using the commands you list) with another error complaining that it requires [BuildKit](https://docs.docker.com/develop/develop-images/build_enhancements/#to-enable-buildkit-builds). This appears to be mitigated by using the (more modern?) `docker compose` version of the command, which appears to work without issue for me.

------

First error:

```
[ … snip … ]
 => [stage-0 5/7] COPY requirements.txt /src/                              0.7s
 => ERROR [stage-0 6/7] RUN --mount=type=cache,target=/root/.cache/pip p  24.2s
------
 > [stage-0 6/7] RUN --mount=type=cache,target=/root/.cache/pip pip install -r requirements.txt:
#14 1.650 Collecting django~=1.11.0 (from -r requirements.txt (line 1))
#14 2.023   Downloading https://files.pythonhosted.org/packages/49/49/178daa8725d29c475216259eb19e90b2aa0b8c0431af8c7e9b490ae6481d/Django-1.11.29-py2.py3-none-any.whl (6.9MB)
#14 4.463 Collecting redis==2.10.6 (from -r requirements.txt (line 2))
#14 4.805   Downloading https://files.pythonhosted.org/packages/3b/f6/7a76333cf0b9251ecf49efff635015171843d9b977e4ffcf59f9c4428052/redis-2.10.6-py2.py3-none-any.whl (64kB)
#14 4.846 Collecting boto==2.49.0 (from -r requirements.txt (line 3))
#14 5.094   Downloading https://files.pythonhosted.org/packages/23/10/c0b78c27298029e4454a472a1919bde20cb182dab1662cec7f2ca1dcc523/boto-2.49.0-py2.py3-none-any.whl (1.4MB)
#14 5.413 Collecting gunicorn==19.3.0 (from -r requirements.txt (line 4))
#14 5.798   Downloading https://files.pythonhosted.org/packages/80/1b/25eaf834e05c82233568cdada4a3ea373fd2bc27c447554a98e18def5494/gunicorn-19.3.0-py2.py3-none-any.whl (110kB)
#14 5.836 Collecting pyScss==1.3.4 (from -r requirements.txt (line 5))
#14 6.321   Downloading https://files.pythonhosted.org/packages/1d/4a/221ae7561c8f51c4f28b2b172366ccd0820b14bb947350df82428dfce381/pyScss-1.3.4.tar.gz (120kB)
#14 6.481 Collecting eventlet~=0.20.1 (from -r requirements.txt (line 6))
#14 6.600   Downloading https://files.pythonhosted.org/packages/5b/3a/a72c3c03382c4c32bfafdb5cf51979d066caa496c8e9d8ecbb9f9153bda8/eventlet-0.20.1-py2.py3-none-any.whl (387kB)
#14 6.801 Collecting pytz (from django~=1.11.0->-r requirements.txt (line 1))
#14 7.171   Downloading https://files.pythonhosted.org/packages/60/2e/dec1cc18c51b8df33c7c4d0a321b084cf38e1733b98f9d15018880fb4970/pytz-2022.1-py2.py3-none-any.whl (503kB)
#14 7.324 Collecting six (from pyScss==1.3.4->-r requirements.txt (line 5))
#14 7.442   Downloading https://files.pythonhosted.org/packages/d9/5a/e7c31adbe875f2abbb91bd84cf2dc52d792b5a01506781dbcf25c91daf11/six-1.16.0-py2.py3-none-any.whl
#14 7.454 Collecting enum34 (from pyScss==1.3.4->-r requirements.txt (line 5))
#14 7.634   Downloading https://files.pythonhosted.org/packages/6f/2c/a9386903ece2ea85e9807e0e062174dc26fdce8b05f216d00491be29fad5/enum34-1.1.10-py2-none-any.whl
#14 7.647 Collecting pathlib (from pyScss==1.3.4->-r requirements.txt (line 5))
#14 7.822   Downloading https://files.pythonhosted.org/packages/ac/aa/9b065a76b9af472437a0059f77e8f962fe350438b927cb80184c32f075eb/pathlib-1.0.1.tar.gz (49kB)
#14 7.938 Collecting enum-compat (from eventlet~=0.20.1->-r requirements.txt (line 6))
#14 8.022   Downloading https://files.pythonhosted.org/packages/41/46/8ed2368976d934652d8f33f1fdd86f5580fab45e890c90a848e83097a093/enum-compat-0.0.3.tar.gz
#14 8.130 Collecting greenlet>=0.3 (from eventlet~=0.20.1->-r requirements.txt (line 6))
#14 23.29   Could not find a version that satisfies the requirement greenlet>=0.3 (from eventlet~=0.20.1->-r requirements.txt (line 6)) (from versions: )
#14 23.50 No matching distribution found for greenlet>=0.3 (from eventlet~=0.20.1->-r requirements.txt (line 6))
#14 23.74 You are using pip version 8.1.1, however version 22.0.4 is available.
#14 23.74 You should consider upgrading via the 'pip install --upgrade pip' command.
------
failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c pip install -r requirements.txt]: exit code: 1
```

Second error:

```
[ … snip … ]
Step 7/8 : RUN --mount=type=cache,target=/root/.cache/pip pip install -r requirements.txt
ERROR: Service 'website' failed to build : the --mount option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled
```

**Comment by SteveMarshall** [comment]
OK, so I think I've got static assets working (at runtime, at least).

In `backend/stats_porn.py` (yeesh, that name has not aged well), we need to do similar Redis config stuff as you've done in

**Comment by jaylett** [comment]
@SteveMarshall cdn.spacelog.org is just a CNAME to a cloudfront distribution which I think Russ originally set up. However I'm going to suggest that you're doing something wrong, because the live site (which admittedly seems to have problems with the core site but is running the mission sites fine) just serves these things locally with the entire thing pushed through Cloudflare which (possibly) then caches some of it. (And even if it doesn't the live system has varnish inline.)

**Comment by jaylett** [comment]
(Docker may cause us to switch back to static hosted assets, although we almost certainly do need varnish inline for the HTML if possible because some of the pages are non-trivial to render IIRC. If we switch to hosted assets and haven't already we probably want to align to the Django approach which didn't exist when we originally wrote the `artemis` code. I haven't checked in on Django recently, but I used to carry some helpful mixins that made the file management work better for assets.)

**Comment by russss** [comment]
There is always the cost-free "just use Cloudflare" option which probably means we wouldn't need any additional caching.  I don't like Cloudflare but the simplicity benefits and the price being free is pretty attractive.

Using a separate host for static assets is no longer a performance benefit in the age of HTTP2 so there's no problem with serving the static assets from the same hostname now.

**Comment by SteveMarshall** [comment]
@jaylett Actually, I’m not sure it is me: I wonder if it’s the fact that the global live site is down (missions are fine). I didn’t check assets locally for missions, but will check again when I get home.

**Comment by jaylett** [comment]
@SteveMarshall on live, missions load relative to the page, so I don't see why global would work any differently if it were working.

I've rebooted the VM which may bring global back, but it could equally be bitrot between the image and the underlying Xen or whatever.

**Comment by jaylett** [comment]
> There is always the cost-free "just use Cloudflare" option which probably means we wouldn't need any additional caching. I don't like Cloudflare but the simplicity benefits and the price being free is pretty attractive.

Right now we don't have the magic page rule to cache HTML, but we can add that when we switch over.

We should also probably complete the transition to HTTPS by adding HSTS. (IIRC we support HTTPS except that some Amazon images are broken because Amazon at the time, but I'm not sure we even redirect to HTTPS.)

**Comment by SteveMarshall** [comment]
@jaylett @russ Right, I'm back home and pretty sure I'm not doing something wrong, but various things are broken or behave differently depending on where they're run. As I'm seeing it at the moment:

1. `https://spacelog.org` (both `http` and `https`, but no `www`) responds with a (Varnish?) 503 across multiple devices, browsers (including `curl -I`, and internet connections.
2. `www.` (both `http` and `https`) is currently serving Apollo 13, not global.

On my original point about `cdn.` and assets: both missions and global, when run locally, point to `cdn.spacelog.org` for fonts (and only fonts; I imagine this was to try to ensure a cache hit for the font when moving between subdomains), and that URL is hard-coded into the CSS. That URL returns `522 Unknown`, as I said above.

I can't verify what the production global site does because of point 2 above, but the production missions all get their fonts via their own domains (specifically, from `../fonts/` in the CSS), rather than referencing `cdn.` at all (ie differently to how the currently-committed code suggests they should).

I imagine the original reason for the absolute CDN URL was to save users re-downloading the font files from one subdomain to the next (or it's just a holdover from when we served everything from `cdn.`), but given that doesn't appear to be how the production site works, and each version is sub-50k, we can probably not worry too much? (2010-era me is rolling in his grave at the idea that a 50k file is inconsequential…)

If we're all happy with just serving the fonts from the current mission/global (I am), then that problem goes away, but if we want to serve them from the currently-in-code-but-broken `cdn.` URL, then we should work out what's going on with that.

**Comment by jaylett** [comment]
So the reason why HEAD looks different to live is that it has unmerged #257 because no one reviewed it, so it just... sat there for a few years. We should probably get that reviewed and merged, or modified as necessary.

**Comment by SteveMarshall** [comment]
@jaylett Ah, that would explain it. I'll cast an eye over the next few days to see if there's anything obviously broken there and, if not, we can get it merged.

**Comment by SteveMarshall** [comment]
OK, if we merge #257 and rebase this on top of that, that fixes all my issues save [fixing the redis connection in `backend/stats_porn.py` above](https://github.com/Spacelog/Spacelog/pull/259#issuecomment-1104449075), which that diff fixes (unless there's a more consistent/appropriate way to do that).

Once that's fixed, I think we are [GO for (production) static assets](http://apollo13.dev.spacelog.org:8000/00:00:02:02/#log-line-122).

@bfirsh Did you have thoughts/opinions on how to do development asset stuff or build-time indexing?

**Comment by bfirsh** [comment]
For the build-time indexing, I was thinking of writing a shell script that spun up a redis server and did an index, then `RUN` that near the end of the `Dockerfile`.

**Comment by jaylett** [comment]
Merged #257 following @SteveMarshall's review.

On build-time indexing, the idea (presumably) is that we build a redis db (and Xapian db) within the docker container, so @bfirsh's approach sounds right to me. You probably need to cleanly shut down redis to get it to sync the db locally. (This is assuming redis still works roughly as it did 10 years ago, so YMMV.)

**Comment by SteveMarshall** [comment]
@bfirsh How do you want to work on this PR? Are you OK for me to push to it (and rebase to incorporate eg #257) as I find things that resolve issues with the build, or would you rather I raise PRs from it?

**Comment by bfirsh** [comment]
Yes, feel free to push to this branch and/or rebase it!

File: backend/indexer.py (hunk 1 above)
Comment (SteveMarshall) [reply backend/indexer.py:633]: Chef! 👨🏼‍🍳🔥

**Comment by SteveMarshall** [comment]
OK! I've rebased this to include #257 (fixing my earlier CSS/font woes), as well as making stats generation work in the new world (8fee7bb), and putting together a first, somewhat naive implementation of addressing build-time indexing (3fd138f). I'd appreciate some feedback on the latter of those (3fd138f) in particular.

Remaining issues I see before we think about making a production image and deploying it somewhere:

- [ ] Development CSS/asset workflow
- [ ] Add an entry point to the docker container/work out how to run redis + app server in production
- [ ] Search appears to be broken (`SearchError at /search/: Field 'weight' was not indexed for sorting`)

I'd also like to see if we can speed up and simplify local first-run: at the moment, we're building the same image twice with two different names (`spacelog_global` and `spacelog_website`). @bfirsh I don't suppose you know any techniques that would allow us to build once and use in both places? (I've been manually building and pointing both at a `spacelog` image, but that seems… suboptimal.) The ideal situation, as I see it, is that we run `docker compose up`, one container is built, and then used by both services.

**Comment by bfirsh** [comment]
There is no way to do that, I think. The idea is that Docker's build caching will hopefully mean that it's using the same image, despite building it twice. It should just run through the build being like "ah, yes, I've built all this before, I'll just use the existing image".

Maybe that's not working right though if it looks like it's fully building it twice.

**Comment by SteveMarshall** [comment]
Yeah, that was my suspicion. I think build caching's not working as one might hope because it's happening literally at the same time, in parallel, when running a straight `docker compose up`, if you see what I mean?

Oh well!

**Comment by bfirsh** [comment]
Oh, weird. That's new behavior I think. :(

**Comment by bfirsh** [comment]
> I think the solution is going to actually be to just build/run one service that serves both website and global once we're using a different init system.

👍 This was intended for the development environment. In production, we can probably just have one Docker image somehow.

**Comment by SteveMarshall** [comment]
I realise today that I've not pushed a bunch of the things I've done over the last few days. The most recent batch of commits:

- Use s6-overlay to allow us to run multiple services in our containers
- Combines global and website into one container
- Serve dev/live assets appropriately (statically for the latter; updating regularly for the former)
- Use nginx + gunicorn by default in the container, and Django's dev server when using `docker compose`

Things that're left to fix:

- Fix links between sites (`PROJECT_DOMAIN` and `PROJECT_HOME` should be overridable with environment variables)
- Redirect logging nginx and gunicorn logging to stdout and stderr
- Fix search in development (@jaylett, is that doable/worth bothering with?)

I'd appreciate thoughts on all of this, and what else we want to do to productions the container… Do we want to include Varnish, too? Is there stuff we want to tweak about any of the stuff I've done here?

**Comment by jaylett** [comment]
@SteveMarshall I'd suggest not bothering with search in development, and fixing it if anyone needs to work on search in future. And I think we can skip varnish and configure caching in Cloudflare for production.

**Comment by SteveMarshall** [comment]
@simonw Thanks for the tip! ✨

**Comment by SteveMarshall** [comment]
@bfirsh @jaylett I _think_ we might be at a point where it's worth us re-reviewing this to see if we're happy for it to be our production image, and then think about where to host it and all that.

Would one or both of you mind casting an eye over my changes and let me know what you think?

**Comment by bfirsh** [comment]
Ooh, exciting. I will do a review of this when I get a chance.

On the subject of where to put it...

> some hosted service that runs Docker containers (Heroku?)

That didn't age well. 😂

FWIW my favorite modern make-a-Docker-image-run service by far is https://fly.io/ . It may well fit within the free plan.

**Comment by simonw** [comment]
In case it's useful here's how to push the logs on Fly to S3 (figure that out this morning): https://til.simonwillison.net/fly/fly-logs-to-s3

**Comment by bfirsh** [comment]
Just done a quick review of this and broadly looks good to me! Appreciate the clear commits with lots of context. Made for an easy review. :)

When Cloudflare is in front of a Django app, I usually serve static assets with [Whitenoise](http://whitenoise.evans.io/en/stable/). The main problem with gunicorn, as I understand it, is performance. Cloudflare will almost always serve the assets, so this makes it fine to use gunicorn.

This reduces the complexity of having to run and configure nginx, etc. [It is the recommended method for deploying on Heroku.](https://devcenter.heroku.com/articles/django-assets)

**Comment by bfirsh** [comment]
I had a go at getting it running on fly.io. One fun problem: it's impossible to run as PID 1, so S6 doesn't work. [Fly recommend overmind](https://fly.io/docs/app-guides/multiple-processes/), which actually looks rather more elegant.

I have invited you both to the fly.io org I'm playing around in, in case you want to give it a try...

**Comment by SteveMarshall** [comment]
Thanks @bfirsh, that’s really useful feedback!

I’ve got a first cut of doing everything using overmind running locally, and will try to incorporate all your feedback (including simplifying actually running everything) over the next couple of days 🤞🏼

**Comment by SteveMarshall** [comment]
So! One… fun… thing to note about WhiteNoise is that it requires Python 3.7 (we're on 2.7.12) and Django 2.2-4.0 (we're on 1.11.0). Sooooo… that's nice 😱🙈

I guess I'll be looking into upgrading Django by 3 versions next, then!

**Comment by SteveMarshall** [comment]
Apologies for the radio silence again: life's been happening 😅

Anyway! I've got it running, after a fashion*, [on fly.io](https://spacelog.fly.dev) (though it's likely 500ing when you read this), but Redis gets killed with an out of memory error after a small handful of requests. I'm reproducing that locally if I limit the container's memory to 100MB, and will see if we can do anything to get Redis to be less of a memory hog 🤔 Suggestions on a postcard (/comment here), if you have them!

* Not pointing domains at it, so it's defaulting to Apollo 13, and I haven't worked out if/how to route traffic in fly based on domain.

**Comment by bfirsh** [comment]
Hmm. I wonder if Cloudflare could do that kind of routing for us. Would probably work on their free plan.

**Comment by SteveMarshall** [comment]
👋🏼 Now I've got DNS access (thanks @jaylett!), I've been able to get fly to correctly serve everything.

You can see it in action (red/blue banner tells you it's served from fly for now) on [Vostok 1](https://vostok1.spacelog.org), [Gemini 3](https://gemini3.spacelog.org/search/?q=One-Alpha), and [global](https://spacelog.fly.dev).

I _think_ that means the only things left to do for this PR are:

- [ ] Remove the red/blue banners and redeploy
- [ ] Update the documentation (about to be in progress)
- [ ] Re-point Cloudflare at fly once everything's merged

After all that's done, and we're live on fly, I'll look at (not necessarily in this order):

- Automating [continuous deployments](https://fly.io/docs/app-guides/continuous-deployment-with-github-actions/) on merge to main
- Removing cruft we no longer need (eg Chef, Vagrant, Fabric, Screen, etc)
- Upgrading everything to latest, removing hacks to enable old versions, and enabling Dependabot
- … and then maybe some of the outstanding issues!

**Comment by SteveMarshall** [comment]
… and I've had a go at adding some documentation, lemme know what you think!

**Review by SteveMarshall — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -50,103 +48,50 @@ Clone the repository from git:
...
-of running multiple programs on one terminal. Currently it leaves you
-looking at the development server log for the global homepage, but you
-can switch to a blank terminal by typing `^A 0`, ie: holding down the
-`ctrl` key, pressing `A`, releasing `ctrl` and then pressing `0`. This
-is a good place to run `make reindex` from.
-
-All our `make` commands will take care of the virtualenv for you; if
-you're not using one, you can use `PYTHON=python make <whatever>`
-instead.
+That will build the Docker container, containing all the current mission content, with CSS and code re-generated as it changes. Once that command finishes (it won't exit, but it will say things like “Starting development server”), you can point your web browser at [http://dev.spacelog.org:8000/](http://dev.spacelog.org:8000/) and the global homepage should come up; from there you can navigate to other missions, which will appear at URLs such as [http://apollo11.dev.spacelog.org:8001/](http://apollo11.dev.spacelog.org:8001). The DNS is managed by us, and providing you're online everything will just work.
```
Comment (SteveMarshall) [inline README.md:74]: The ports in all of these links are wrong.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# syntax = docker/dockerfile:1.2
```
[the remark this answers is not in the document]
Comment (SteveMarshall) [reply Dockerfile:1]: There's not _that_ much difference either way, so I think let's just leave it for now.

**Comment by SteveMarshall** [comment]
@jaylett @bfirsh @russss Right! I _think_ this is ready for one or all of you to cast an eye, if you wouldn't mind? Once some of you are happy with it, I'll do the deploying/DNS updating dances we need to do and get this merged.

**Review by SteveMarshall — DISMISSED** [review]

**Comment by SteveMarshall** [comment]
@jaylett @bfirsh @russss The fly-hosted version has been running happily for a little over 6 months, and the old version has been down for almost that entire time. With that in mind, I've removed the coloured banners from the fly version and re-pointed all the DNS (including adding IPv6 records) at that. It's holding up nicely at the moment.

So: if I don't hear from any of you by… let's say 5pm BST next Friday 22nd March, I'm going to call this done and merge it so that I can move on to some of the other things. Hopefully that's OK; let me know if you've any issues with it, though, or would like more time to review!

**Comment by SteveMarshall** [comment]
OK, as there have been no comments, and these changes are live and have been happily ticking away in Fly for months (for Vostok 1, Gemini 3, and global, anyway), I'm going to get this merged and move on to other things.

Thanks for your input, all!

**Comment by bfirsh** [comment]
Hooray! Thanks for getting this over the line.

**Comment by SteveMarshall** [comment]
@bfirsh My pleasure! Thank you for kicking it off and for your guidance 👍🏼

**Pull request merged**